Out-of-tree docs build fails due to missing remote

I assume this is due to me doing an out-of-tree build

```
make O=$HOME/builds/julia configure
cd $HOME/builds/julia
make -j
make docs
```

```
┌ Debug: Rendering stdlib/Downloads.md [68]
└ @ Documenter.HTMLWriter ~/src/julia/doc/deps/packages/Documenter/1HwWe/src/html/HTMLWriter.jl:750
ERROR: LoadError: MissingRemoteError: unable to generate source url
  path: /home/vchuravy/src/julia/doc/src/stdlib/Downloads.md
Documenter was unable to automatically determine the remote repository for this file.
This can happen if you are including docstrings or pages from secondary packages. Those packages
must be cloned as Git repositories (i.e. Pkg.develop instead Pkg.add), or the `remotes` keyword
must be configured appropriately. See the 'Remote repository links' section in the manual for
more information.
Stacktrace:
  [1] edit_url(doc::Documenter.Document, path::String; rev::String)
    @ Documenter ~/src/julia/doc/deps/packages/Documenter/1HwWe/src/documents.jl:866
  [2] edit_link(f::Documenter.HTMLWriter.var"#51#53"{Documenter.DOM.Node, Documenter.DOM.Tag, Documenter.DOM.Tag}, ctx::Documenter.HTMLWriter.HTMLContext, navnode::Documenter.NavNode)
    @ Documenter.HTMLWriter ~/src/julia/doc/deps/packages/Documenter/1HwWe/src/html/HTMLWriter.jl:1338
  [3] render_navbar(ctx::Documenter.HTMLWriter.HTMLContext, navnode::Documenter.NavNode, edit_page_link::Bool)
    @ Documenter.HTMLWriter ~/src/julia/doc/deps/packages/Documenter/1HwWe/src/html/HTMLWriter.jl:1269
  [4] render_page(ctx::Documenter.HTMLWriter.HTMLContext, navnode::Documenter.NavNode)
    @ Documenter.HTMLWriter ~/src/julia/doc/deps/packages/Documenter/1HwWe/src/html/HTMLWriter.jl:841
  [5] (::Documenter.HTMLWriter.var"#18#26"{Documenter.Document, Documenter.HTMLWriter.HTMLContext})(page::String)
    @ Documenter.HTMLWriter ~/src/julia/doc/deps/packages/Documenter/1HwWe/src/html/HTMLWriter.jl:751
  [6] iterate
    @ ./generator.jl:48 [inlined]
  [7] collect_to!
    @ ./array.jl:834 [inlined]
  [8] collect_to_with_first!
    @ ./array.jl:812 [inlined]
  [9] _collect(c::Vector{String}, itr::Base.Generator{Vector{String}, Documenter.HTMLWriter.var"#18#26"{Documenter.Document, Documenter.HTMLWriter.HTMLContext}}, ::Base.EltypeUnknown, isz::Base.HasShape{1})
    @ Base ./array.jl:806
 [10] collect_similar(cont::Vector{String}, itr::Base.Generator{Vector{String}, Documenter.HTMLWriter.var"#18#26"{Documenter.Document, Documenter.HTMLWriter.HTMLContext}})
    @ Base ./array.jl:705
 [11] map
    @ ./abstractarray.jl:3295 [inlined]
 [12] render(doc::Documenter.Document, settings::Documenter.HTMLWriter.HTML)
    @ Documenter.HTMLWriter ~/src/julia/doc/deps/packages/Documenter/1HwWe/src/html/HTMLWriter.jl:747
 [13] runner(::Type{Documenter.HTMLFormat}, fmt::Documenter.HTMLWriter.HTML, doc::Documenter.Document)
    @ Documenter ~/src/julia/doc/deps/packages/Documenter/1HwWe/src/writers.jl:23
 [14] dispatch(::Type{Documenter.FormatSelector}, ::Documenter.HTMLWriter.HTML, ::Vararg{Any})
    @ Documenter.Selectors ~/src/julia/doc/deps/packages/Documenter/1HwWe/src/utilities/Selectors.jl:170
 [15] render(doc::Documenter.Document)
    @ Documenter ~/src/julia/doc/deps/packages/Documenter/1HwWe/src/writers.jl:45
 [16] runner(::Type{Documenter.Builder.RenderDocument}, doc::Documenter.Document)
    @ Documenter ~/src/julia/doc/deps/packages/Documenter/1HwWe/src/builder_pipeline.jl:258
 [17] dispatch(::Type{Documenter.Builder.DocumentPipeline}, x::Documenter.Document)
    @ Documenter.Selectors ~/src/julia/doc/deps/packages/Documenter/1HwWe/src/utilities/Selectors.jl:170
 [18] #83
    @ ~/src/julia/doc/deps/packages/Documenter/1HwWe/src/makedocs.jl:248 [inlined]
 [19] withenv(::Documenter.var"#83#85"{Documenter.Document}, ::Pair{String, Nothing}, ::Vararg{Pair{String, Nothing}})
    @ Base ./env.jl:257
 [20] #82
    @ ~/src/julia/doc/deps/packages/Documenter/1HwWe/src/makedocs.jl:247 [inlined]
 [21] cd(f::Documenter.var"#82#84"{Documenter.Document}, dir::String)
    @ Base.Filesystem ./file.jl:112
 [22] makedocs(; debug::Bool, format::Documenter.HTMLWriter.HTML, kwargs::@Kwargs{build::String, modules::Vector{Module}, clean::Bool, doctest::Bool, linkcheck::Bool, linkcheck_ignore::Vector{String}, checkdocs::Symbol, sitename::String, authors::String, pages::Vector{Any}, remotes::Dict{String, Tuple{Documenter.Remotes.GitHub, String}}})
    @ Documenter ~/src/julia/doc/deps/packages/Documenter/1HwWe/src/makedocs.jl:247
 [23] top-level scope
    @ ~/src/julia/doc/make.jl:371
in expression starting at /home/vchuravy/src/julia/doc/make.jl:371
make[1]: *** [/home/vchuravy/src/julia/doc/Makefile:47: html] Error 1
make: *** [/home/vchuravy/src/julia/Makefile:120: docs] Error 2
```

--- a/doc/Makefile
+++ b/doc/Makefile
@@ -30,6 +30,13 @@ $(SRCCACHE)/UnicodeData-$(UNICODE_DATA_VERSION).txt:
 	@mkdir -p "$(SRCCACHE)"
 	$(JLDOWNLOAD) "$@" https://www.unicode.org/Public/$(UNICODE_DATA_VERSION)/ucd/UnicodeData.txt
 
+# NEWS.md and stdlib are in-tree build artifacts - don't link them for oot builds.
+DOC_FILES=$(filter-out NEWS.md stdlib,$(notdir $(wildcard $(SRCDIR)/src/*)))
+src/%:
+	@mkdir -p src
+	ln -s $(SRCDIR)/src/$* $@
+src: $(addprefix src/,$(DOC_FILES))
+
 deps: $(SRCCACHE)/UnicodeData-$(UNICODE_DATA_VERSION).txt
 	$(JLCHECKSUM) "$<"
 	cp "$<" UnicodeData.txt

--- a/doc/make.jl
+++ b/doc/make.jl
@@ -1,9 +1,13 @@
+let r = r"buildroot=(.+)", i = findfirst(x -> occursin(r, x), ARGS)
+    global const buildrootdoc = i === nothing ? (@__DIR__) : joinpath(first(match(r, ARGS[i]).captures), "doc")
+end
+
 # Install dependencies needed to build the documentation.
 Base.ACTIVE_PROJECT[] = nothing
 empty!(LOAD_PATH)
 push!(LOAD_PATH, @__DIR__, "@stdlib")
 empty!(DEPOT_PATH)
-push!(DEPOT_PATH, joinpath(@__DIR__, "deps"))
+push!(DEPOT_PATH, joinpath(buildrootdoc, "deps"))
 push!(DEPOT_PATH, abspath(Sys.BINDIR, "..", "share", "julia"))
 using Pkg
 Pkg.instantiate()
@@ -26,7 +30,7 @@ cp_q(src, dest) = isfile(dest) || cp(src, dest)
 const STDLIB_DOCS = []
 const STDLIB_DIR = Sys.STDLIB
 const EXT_STDLIB_DOCS = ["Pkg"]
-cd(joinpath(@__DIR__, "src")) do
+cd(joinpath(buildrootdoc, "src")) do
     Base.rm("stdlib"; recursive=true, force=true)
     mkdir("stdlib")
     for dir in readdir(STDLIB_DIR)
@@ -70,7 +74,8 @@ function parse_stdlib_version_file(path)
 end
 # This generates the value that will be passed to the `remotes` argument of makedocs(),
 # by looking through all *.version files in stdlib/.
-documenter_stdlib_remotes = let stdlib_dir = realpath(joinpath(@__DIR__, "..", "stdlib"))
+documenter_stdlib_remotes = let stdlib_dir = realpath(joinpath(@__DIR__, "..", "stdlib")),
+                                stdlib_build_dir = joinpath(buildrootdoc, "..", "stdlib")
     # Get a list of all *.version files in stdlib/..
     version_files = filter(readdir(stdlib_dir)) do fname
         isfile(joinpath(stdlib_dir, fname)) && endswith(fname, ".version")
@@ -97,7 +102,7 @@ documenter_stdlib_remotes = let stdlib_dir = realpath(joinpath(@__DIR__, "..", "
             versionfile[sha_key]
         end
         # Construct the absolute (local) path to the stdlib package's root directory
-        package_root_dir = joinpath(stdlib_dir, "$(package)-$(package_sha)")
+        package_root_dir = joinpath(stdlib_build_dir, "$(package)-$(package_sha)")
         # Documenter needs package_root_dir to exist --- it's just a sanity check it does on the remotes= keyword.
         # In normal (local) builds, this will be the case, since the Makefiles will have unpacked the standard
         # libraries. However, on CI we do this thing where we actually build docs in a clean worktree, just
@@ -127,7 +132,7 @@ function generate_markdown(basename)
     @assert length(splitted) == 2
     replaced_links = replace(splitted[1], r"\[\#([0-9]*?)\]" => s"[#\g<1>](https://github.com/JuliaLang/julia/issues/\g<1>)")
     write(
-        joinpath(@__DIR__, "src", "$basename.md"),
+        joinpath(buildrootdoc, "src", "$basename.md"),
         """
         ```@meta
         EditURL = "https://github.com/JuliaLang/julia/blob/master/$basename.md"
@@ -281,7 +286,7 @@ end
 
 const use_revise = "revise=true" in ARGS
 if use_revise
-    let revise_env = joinpath(@__DIR__, "deps", "revise")
+    let revise_env = joinpath(buildrootdoc, "deps", "revise")
         Pkg.activate(revise_env)
         Pkg.add("Revise"; preserve=Pkg.PRESERVE_NONE)
         Base.ACTIVE_PROJECT[] = nothing
@@ -351,10 +356,6 @@ DocMeta.setdocmeta!(
     recursive=true, warn=false,
 )
 
-let r = r"buildroot=(.+)", i = findfirst(x -> occursin(r, x), ARGS)
-    global const buildroot = i === nothing ? (@__DIR__) : first(match(r, ARGS[i]).captures)
-end
-
 const format = if render_pdf
     Documenter.LaTeX(
         platform = "texplatform=docker" in ARGS ? "docker" : "native"
@@ -377,8 +378,9 @@ else
     )
 end
 
-const output_path = joinpath(buildroot, "doc", "_build", (render_pdf ? "pdf" : "html"), "en")
+const output_path = joinpath(buildrootdoc, "_build", (render_pdf ? "pdf" : "html"), "en")
 makedocs(
+    source    = joinpath(buildrootdoc, "src"),
     build     = output_path,
     modules   = [Main, Base, Core, [Base.root_module(Base, stdlib.stdlib) for stdlib in STDLIB_DOCS]...],
     clean     = true,
@@ -479,7 +481,7 @@ if "deploy" in ARGS
     deploydocs(
         repo = "github.com/JuliaLang/docs.julialang.org.git",
         deploy_config = BuildBotConfig(),
-        target = joinpath(buildroot, "doc", "_build", "html", "en"),
+        target = joinpath(buildrootdoc, "_build", "html", "en"),
         dirname = "en",
         devurl = devurl,
         versions = Versions(["v#.#", devurl => devurl]),

